### DISCUSSION TITLE
Delta snapshot

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#219] [Q&A] [answered]
While working on networking, I got curious if Jolt has a support for delta snapshots (similar to Source Engine from Valve), where subsequent snapshots are smaller due to containing only the changes from the last snapshot. Could be a feature request, if there is no support yet.

**Comment by jrouwe** [comment, marked as answer]
The short answer is no, there isn't. Jolt does expose a `SaveState` function on a per Body level that could be utilized. This would involve looping over all bodies, calling `SaveState`, comparing the stream (hash?) with the previous frame and sending the data if it changed. Doing it like this would send over positions/velocities etc. You would not be a sending over the state for a fully deterministic simulation (I'm not sure if you're looking for that).

**Reply by LeXXik** [reply]
Right, I would need it to be deterministic, so, I guess it is a "no" then. I will use what is available. Thank you for the SaveState tip though, I will keep it in mind when I don't need determinism.

**Reply by jrouwe** [reply]
If you would need it deterministic*, then you would need to call `SaveState` on the Constraints as well, you could use the same trick to decide if data needs to be sent or not. Finally you would need to call `SaveState(.., EStateRecorderState_Global | EStateRecorderState_Contacts, ...)` on `PhysicsSystem` to store the contacts. Delta compression is not possible there (although I'm sure you could implement an algorithm to diff the binary streams).

\* I don't know if the JS version of Jolt is deterministic across platforms/browsers, the C++ version is

**Reply by LeXXik** [reply]
> If you would need it deterministic*, then you would need to call SaveState on the Constraints as well

Ah, I see. Hmm, then perhaps that could well be an option.

> * I don't know if the JS version of Jolt is deterministic across platforms/browsers, the C++ version is

Well, if C++ is deterministic and Wasm instances on different platforms/browsers get the same input in the same order, then it will be deterministic.

**Comment by LeXXik** [comment]
Hmm, do I understand correctly that virtual characters are not included in the recorded state? The state seem to only restore the rigidbodies.

**Reply by jrouwe** [reply]
The `PhysicsSystem` does not keep track of `CharacterVirtual` objects (you're the sole owner), so it has no way of saving their state. You need to call the `CharacterVirtual::SaveState` function yourself.

**Reply by jrouwe** [reply]
See https://github.com/jrouwe/JoltPhysics.js/commit/873a688c1afa78e9740870346118ae2df27f65d3
